File: LinkedLists/Python 3/SingleLinkedList/LinkedListWrap.py
# -*- coding: utf-8 -*-
"""
Created on Sat Jun 23 08:54:12 2018

@author: Sakshi Panday

@Overview Wrapper Class for Linked List
"""
from Node import node

class LinkedListWrap(object):
    def __init__(self):
        self.head = None
        
    def setHead(self, headNode):
        self.head = headNode
        
    def getHead(self):
        return self.head
    
    '''
    This function adds a new node with the data passed to the single linked list
    '''
    def addNode(self, data):
        if self.getHead() == None:
            self.setHead(node(data))
        else:
            currentNode = self.getHead()
            while(currentNode.getNext() != None):
                currentNode = currentNode.getNext()
            currentNode.setNext(node(data))
        return "Added a new node:" + str(data)
 
    '''
    This function deletes the node(s), with data same as value passed, from the single linked list
    '''    
    def deleteNodeByValue(self, value):
        response = ""
        foundCount = 0
        if self.getHead() == None:
            response = "Linked List is Empty!"
        else:
            currentNode = self.getHead()
            if currentNode.getData() == value and currentNode.getNext() == None:
                    self.setHead(self.getHead().getNext())
                    currentNode = self.getHead()
                    foundCount += 1
            else:
                while(currentNode.getNext() != None):
                    if currentNode.getData() == value and currentNode == self.getHead():
                        self.setHead(self.getHead().getNext())
                        currentNode = self.getHead()
                        foundCount += 1
                    elif currentNode.getNext().getData() == value:
                        foundCount += 1
                        currentNode.setNext(currentNode.getNext().getNext())
                    else:
                        currentNode = currentNode.getNext()
            response = "Number of nodes with value '" + str(value) + "' deleted: " + str(foundCount)
        return response

# =============================================================================
#   Deleting the last node of a linked list 
# =============================================================================
    def deleteLastNode(self):
        if self.getHead() == None:
            return "Linked List is Empty!"
        else:
            currentNode = self.getHead()
            if(currentNode.getNext() == None ):
                self.setHead(None)
            else:
                while ( currentNode.getNext().getNext() != None):
                    currentNode = currentNode.getNext()
                currentNode.setNext(None)
        return("Linked List after deletion of last node:\n" + self.__repr__())
        
# =============================================================================
#     Reversing a Single Linked List        
# =============================================================================
    def reverseSingleLL(self):
        if self.getHead()==None:
            return "Linked List is Empty!"
        elif self.getHead().getNext() == None:
            return "Reversed Linked List :\n" + self.__str__()
        else:
            tempList = LinkedListWrap()
            while(self.getHead() != None):
                currentNode = self.getHead()
                while(currentNode.getNext() != None):
                    currentNode = currentNode.getNext()
                tempList.addNode(currentNode.getData())
                self.deleteLastNode()
            self.setHead(tempList.getHead())
            return "Reversed Linked List :\n" + self.__repr__()
        
    def __repr__(self):
        if self.getHead() == None:
            return "Linked List is Empty!"
        currentNode = self.getHead()
        output = str(currentNode.getData())
        while(currentNode.nextLink != None):
            output += ' -> ' + str(currentNode.getNext().getData())
            currentNode = currentNode.getNext()
        return output
    
    def __str__(self):
        currentNode = self.getHead()
        output = str(currentNode.getData())
        while(currentNode.nextLink != None):
            output += ' ' + str(currentNode.getNext().getData())
            currentNode = currentNode.getNext()
        return output

File: LinkedLists/JAVA/LinkedList/src/SingleLinkedList/LinkedListWrap.java
/**
 * 
 */
package SingleLinkedList;

/**
 * @author Sakshi Panday Jun 21, 2018
 */
public class LinkedListWrap {

	Node head;

	/**
	 * constructor for wrapper class of Single Linked List
	 * 
	 */
	public LinkedListWrap() {
		setHead(null);
	}

	/**
	 * This function is used to add a new node at the end of the single linked list
	 * 
	 * @param value
	 */
	public void addNewNode(String value) {
		if (head == null)
			setHead(new Node(value)); // if it's the first node, we need to set it as the new head
		else {
			Node currentNode = head;
			while (currentNode.next != null)
				currentNode = currentNode.next;
			currentNode.setNext(new Node(value)); // and node is added next to the previous node as single directed list
		}
		System.out.println("New node added: " + value);
	}

	/**
	 * This function is used to delete a node of the single linked list by value
	 * 
	 * @param value
	 * @return
	 */
	public String deleteNodeByValue(String value) {
		if (getHead() == null)
			return "Single Linked List is EMPTY";
		Node currentNode = getHead();
		int foundCount = 0;
		if (currentNode.getData() == value && currentNode.equals(getHead())) {
			foundCount += 1;
			this.setHead(currentNode.getNext()); // if the first node is deleted, we need to set the new head to the
			currentNode = getHead(); // next node
		}
		while (currentNode != null && currentNode.getNext() != null) {
			if (currentNode.getData() == value && currentNode.equals(getHead())) {
				foundCount += 1;
				this.setHead(currentNode.getNext()); // if the first node is deleted, we need to set the new head to the
				currentNode = getHead(); // next node
			} else if (currentNode.getNext().getData() == value) {
				foundCount += 1;
				currentNode.setNext(currentNode.getNext().getNext());// if node is deleted, current node's next node is
																		// still unexplored, so no alteration needed for
																		// while loop
			} else
				currentNode = currentNode.getNext(); // if no changes made, iteration for while loop needed
		}
		if (foundCount == 0)
			return "No node is same as the input value in the single linked list";
		return "Number of Nodes with Value '" + value + "' Deleted: " + foundCount;
	}

	@Override
	public String toString() {
		String llInfo = "";
		if (head == null)
			return "Current Single Linked List: EMPTY";
		else {
			Node currentNode = getHead();
			while (currentNode.next != null) {
				llInfo = llInfo.concat(currentNode.data);
				llInfo = llInfo.concat(" ");
				currentNode = currentNode.next;
			}
			llInfo = llInfo.concat(currentNode.data);
		}
		return "Current Single Linked List:\n" + llInfo + "\n";
	}

	public Node getHead() {
		return head;
	}

	public void setHead(Node head) {
		this.head = head;
	}

}
